### PR TITLE
[TensorExpr] IRSimplifier: sort terms in polynomials, terms, minterms, maxterms.

### PR DESCRIPTION
**Pull request by ZolotukhinM**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #63216
* __->__ #63197
* #63196

This solves non-determinism from using hash values in sort methods.
Changes in tests are mostly mechanical.

Differential Revision: [D30292776](https://our.internmc.facebook.com/intern/diff/D30292776)

### CODE REVIEW

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 1]
```diff
@@ -52,8 +59,15 @@ void Polynomial::sort() {
   if (dtype().is_floating_point()) {
     throw std::logic_error("reordering FP ops");
   }
+  std::unordered_map<ExprPtr, std::string> str_repr_cache;
   std::sort(variables_.begin(), variables_.end(), [&](ExprPtr a, ExprPtr b) {
-    return hasher_.hash(a) < hasher_.hash(b);
+    if (!str_repr_cache.count(a)) {
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:64]: This seems to be useful enough to have it as a method. Just a suggestion. I'll leave it to you though.

File: torch/csrc/jit/tensorexpr/ir_simplifier.cpp [hunk 2]
```diff
@@ -2822,6 +2872,49 @@ bool exprEquals(ExprPtr A, ExprPtr B) {
   }
 }
 
+ExprPtr IRSimplifier::simplify(ExprPtr e) {
+  GRAPH_DEBUG("(Simplifier) Original: ", std::to_string(e));
```
Comment (navahgar) [inline torch/csrc/jit/tensorexpr/ir_simplifier.cpp:2876]: Thanks for adding these. I've been wanting to add this to quickly check if simplifier is correct (given the complicated bugs we have in this).

**Review by navahgar — APPROVED** [review]
LGTM

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Comment by ZolotukhinM** [comment]
@ZolotukhinM has imported this pull request.  If you are a Facebook employee, you can view this diff [on Phabricator](https://www.internalfb.com/diff/D30292776).

**Pull request closed without merging**